Make the polling delay dynamic

FichteFoll opened the pull request:
Since the previous implementation uses a single poll-based task per view with a fixed sleep time, we could run into the situation of a poll inside `debounce` happening just before another invocation of the event hook. In that case, we would wait for the full delay time, poll again, notice that the handler has been called and enter another iteration of sleeping and polling.

We can circumvent this by recording a timestamp of the last handler invocation and then calculating the next poll to happen right after the specified delay, relative to the latest invocation.

Disclaimer: I did not test this (because I really *am* that lazy …).

FichteFoll committed:
Make the polling delay dynamic

Since the previous implementation uses a single poll-based task with a
fixed sleep time, we could run into the situation of a poll inside
`debounce` happening just before another invocation of the event hook.
In that case, we would wait for the full delay time, poll again, notice
that the handler has been called and enter another iteration of
sleeping and polling.

We can circumvent this by recording a timestamp of the last handler
invocation and then calculating the next poll to happen right after the
specified delay, relative to the latest invocation.

deathaxe commented on src/sublime_aio.py:
```diff
@@ -82,6 +83,24 @@ def shutdown():
+    """
+    if not __loop:
+        return
+    asyncio.run_coroutine_threadsafe(lock.acquire(), __loop)
```
With regards to the overhead `run_coroutine_threadsafe` causes, I don't like the idea of calling itmore than absolutely needed, especially in possibly heavy duty functions like a debouncer.

FichteFoll replied:
Dropped the lock commit so it can be discussed separately.

deathaxe commented:
Did you face any practical issues with current situation? Otherwise I don't think we need to be overeager with regards to debouncing. One cycle more or less is often not of any importance, IMHO.

FichteFoll commented:
> Did you face any practical issues with current situation?

No, I did not (because I have actually never *used* `sublime_aio`), but I did have experiences like this with other related algorithms.

Regardless, I can easily construct one. It will mostly surface when using higher delays, e.g. 1s, since lower delays likely won't be noticed. When using such an 1s debounce delay on an `on_modified` handler and then a character roughly every second, it is likely that the real function will be called 2s after the last input, which is quite noticeable compared to just 1s. If you are more trained, I'm certain you can also notice it with an original delay of 500ms.

That aside, I don't find this change adding much complexity compared to the better debounce behavior we will get from it.

deathaxe committed:
Some ideas

1. directly import `time.monolitic as now` as shorter alternative to creating
   intermediate `now` variables in debouncer.

2. reorder `time_to_wait` expression to start with static delay followed by
   dynamic part (diff between event trigger and now())

3. call `del call_at` in debounce() directly after checking escape condition
   to group `call_at` access related expressions

4. drop intermediate `view.view_id` in debounce() as it is called way less
   frequently then in wrapper() and thus isn't probably worth it.

5. drop intermediate `pending` variable

Step 3 and 5 prepare for better thread safety measures.

deathaxe commented:
Well, with such debouncers being used to delay input to call linters or update diagnostics etc. this is exactly the kind of extra delay actually really don't care about or find it acceptable. Be it somewhat noticible or not. Delay jitter in delay caused by execution time of especially external processes to create regions is way higher.
